ecl/hql/hqlattr.cpp

```diff
@@ -3776,6 +3776,71 @@ ITypeInfo * setStreamedAttr(ITypeInfo * _type, bool setValue)
     }
 }
 
+//---------------------------------------------------------------------------------
+double queryLikelihood(IHqlExpression * expr)
+{
+    IInterface * match = meta.queryExistingProperty(expr, EPlikelihood);
+    if (match)
+    {
+        IHqlExpression * likelihoodExpr = static_cast<IHqlExpression *>(match);
+        IValue * value = likelihoodExpr->queryChild(0)->queryValue();
+        return value->getRealValue();
+    }
+
+    double likelihoodValue = -1.0;
+    switch(expr->getOperator())
+    {
+    case no_likely:
+        if (expr->numChildren() > 1)
+        {
+            IValue *value = expr->queryChild(1)->queryValue();
+            likelihoodValue = value->getRealValue();
+        }
+        else
+            likelihoodValue = 0.99;
+        break;
+    case no_unlikely:
+        likelihoodValue = 0.01;
+        break;
+    case no_alias:
+        return queryLikelihood(expr->queryChild(0));
+    case no_nofold:
+        {
+            IHqlExpression * child = expr->queryChild(0);
+            IValue * boolValue = child->queryValue();
+            if ( !boolValue )
+                break;
+            if (boolValue->getBoolValue())
+                likelihoodValue = 0.99999999;
+            else
+                likelihoodValue = 0.00000001;
+        }
+        break;
+    default:
+        likelihoodValue = -1.0;
+        break;
+    }
+
+    OwnedHqlExpr likelihoodExpr;
+    Owned<IValue> value = createRealValue(likelihoodValue, 8);
+    likelihoodExpr.setown(createConstant(value.getClear()));
+
+    meta.addProperty(expr, EPlikelihood, likelihoodExpr);
+    return likelihoodValue;
```

Comment:
I would have the main function as queryLikelihood(expr) , and return likelyHoodExpr at this point.

Then implement a queryLikelihood() using queryLikelihoodExpr()->queryValue()->getRealValue();